scene/gui/tree.cpp

```diff
@@ -195,6 +195,72 @@ bool TreeItem::is_indeterminate(int p_column) const {
 	return cells[p_column].indeterminate;
 }
 
+void TreeItem::propagate_check(int p_column) {
+	bool ch = cells[p_column].checked;
+
+	_propagate_check_through_children(p_column, ch);
+	_propagate_check_through_parents(p_column);
+}
+
+TypedArray<TreeItem> TreeItem::propagate_check_and_get_affected(int p_column) {
+	TypedArray<TreeItem> affected_items;
+	bool ch = cells[p_column].checked;
+
+	affected_items.append(this);
+
+	_propagate_check_through_children(p_column, ch, &affected_items);
+	_propagate_check_through_parents(p_column, &affected_items);
+	return affected_items;
+}
+
+void TreeItem::_propagate_check_through_children(int p_column, bool p_checked, TypedArray<TreeItem> *p_arr) {
+	TreeItem *current = get_first_child();
+	while (current) {
+		if (p_arr) {
+			p_arr->append(current);
+		}
+		current->set_checked(p_column, p_checked);
+		current->_propagate_check_through_children(p_column, p_checked, p_arr);
+		current = current->get_next();
+	}
+}
+
+void TreeItem::_propagate_check_through_parents(int p_column, TypedArray<TreeItem> *p_arr) {
+	TreeItem *current = get_parent();
+	if (!current) {
+		return;
+	}
+
+	bool all_unchecked_and_not_indeterminate = true;
+	bool any_unchecked_or_indeterminate = false;
+
+	TreeItem *child_item = current->get_first_child();
+	while (child_item) {
+		if (!child_item->is_checked(p_column)) {
+			any_unchecked_or_indeterminate = true;
+			if (child_item->is_indeterminate(p_column)) {
+				all_unchecked_and_not_indeterminate = false;
+			}
+		} else {
+			all_unchecked_and_not_indeterminate = false;
+		}
+		child_item = child_item->get_next();
+	}
```

Comment:
You can break early out of this loop if `all_unchecked_and_not_indeterminate` is false and `any_unchecked_or_indeterminate` is true, as they won't change anymore.